cabot/cabotapp/views.py

```diff
@@ -478,107 +478,77 @@ def get(self, *args, **kwargs):
-        return HttpResponse(self.template.render(c))
+class UserProfileUpdateAlert(LoginRequiredMixin, UpdateView):
+    """
+    The "My Profile" page. Allows editing the current User model and any AlertPluginUserDatas.
+    The 'alerttype' kwarg is the AlertPluginUserData.title we're editing, 'pk' is the user whose preferences we're
+    editing (superuser can edit anyone's settings).
+    The 'General' alerttype is a special case for editing the User model (from Django).
+    """
+    template_name = 'cabotapp/alertpluginuserdata_form.html'
 
-    def post(self, request, pk, alerttype):
-        if request.user.id != int(pk) and not request.user.is_superuser:
+    # return a 403 if not logged in or the user ID doesn't match the logged in user and the user isn't an admin
+    # (this runs before get/post)
+    def dispatch(self, request, *args, **kwargs):
+        if request.user.id != int(kwargs['pk']) and not request.user.is_superuser:
             return HttpResponse(status=403)
+        return super(UserProfileUpdateAlert, self).dispatch(request, *args, **kwargs)
 
-        profile = UserProfile.objects.get(user=pk)
-        if (alerttype == u'General'):
-            form = GeneralSettingsForm(request.POST)
-            if form.is_valid():
-                profile.user.first_name = form.cleaned_data['first_name']
-                profile.user.last_name = form.cleaned_data['last_name']
-                profile.user.is_active = form.cleaned_data['enabled']
-                profile.user.email = form.cleaned_data['email_address']
-                profile.user.save()
-                profile.save()
-                return HttpResponseRedirect(reverse('update-alert-user-data', args=(self.kwargs['pk'], alerttype)))
-
-        else:
-            plugin_userdata = self.model.objects.get(title=alerttype, user=profile)
-            form_model = get_object_form(type(plugin_userdata))
-            form = form_model(request.POST, instance=plugin_userdata)
-            if form.is_valid() and not form.errors:
-                form.save()
-                return HttpResponseRedirect(reverse('update-alert-user-data', args=(self.kwargs['pk'], alerttype)))
-            else:
-                c = RequestContext(request, {
-                    'form': form,
-                    'alert_preferences': profile.user_data,
-                })
-                return HttpResponse(self.template.render(c))
+    def get_object(self, queryset=None):
+        alerttype = self.kwargs['alerttype']
 
+        user = User.objects.get(pk=self.kwargs['pk'])
+        if alerttype == u'General':
+            return user
 
-def get_object_form(model_type):
-    class AlertPreferencesForm(forms.ModelForm):
-        class Meta:
-            model = model_type
+        profile = UserProfile.objects.get_or_create(user=user)[0]
+        profile.user_data()  # side-effect of creating AlertPluginUserData for user
+        return AlertPluginUserData.objects.get(title=alerttype)
```

Comment:
filter by user

Reply:
FYI - found that django's `.get()` actually raises an error when there is more than one user (instead of happily returning the first matching object, which I assumed). So this would've just made the profile page return 500 in a live environment.